C compilation error when attempting to insert a struct value into a array of struct references

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.12, press to see full `v doctor` output</summary>

|V full version      |V 0.4.12 7f6db12a461a5388b1b9765acec74138ce2817bc
|:-------------------|:-------------------
|OS                  |linux, Linux version 6.16.11-200.fc42.x86_64 (mockbuild@2a5103c3cf1c4acbaad71d7113e31c7b) (gcc (GCC) 15.2.1 20250808 (Red Hat 15.2.1-1), GNU ld version 2.44-6.fc42) #1 SMP PREEMPT_DYNAMIC Mon Oct  6 20:00:39 UTC 2025
|Processor           |16 cpus, 64bit, little endian, AMD Ryzen 7 7800X3D 8-Core Processor
|Memory              |15GB/30.94GB
|                    |
|V executable        |/home/yann/Apps/v/v
|V last modified time|2025-10-05 13:36:24
|                    |
|V home dir          |OK, value: /home/yann/Apps/v
|VMODULES            |OK, value: /home/yann/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/yann/devel/perso/v/bug_report
|                    |
|Git version         |git version 2.51.0
|V git status        |N/A
|.git/config present |false
|                    |
|cc version          |cc (GCC) 15.2.1 20250808 (Red Hat 15.2.1-1)
|gcc version         |gcc (GCC) 15.2.1 20250808 (Red Hat 15.2.1-1)
|clang version       |clang version 20.1.8 (Fedora 20.1.8-4.fc42)
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.41

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  array_insert_compile_bug.v && array_insert_compile_bug`
```v
module main

struct Dummy {
	v int
}

@[heap]
struct Container {
mut:
	vals []&Dummy
}

fn (mut c Container) add(v Dummy) {
	c.vals.insert(0, v)
}

fn main() {
	mut c := &Container{}

	d := Dummy{v: 1}
	println(d)

	c.add(d)
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/array_insert_compile_bug.01K7M886K6KFASARFHDD4EYB7W.tmp.c:5225: error: cannot convert 'struct main__Dummy' to 'struct main__Dummy *'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

An V compilation error indicating that the v Dummy argument couldn't not be inserted into an []&Dummy array.


> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -3321,8 +3321,15 @@ fn (mut c Checker) array_builtin_method_call(mut node ast.CallExpr, left_type as
 				node.args[i].typ = c.expr(mut arg.expr)
 				if i == val_arg_n {
 					arg_sym := c.table.sym(node.args[i].typ)
-					if !c.check_types(node.args[i].typ, info.elem_type)
-						&& !c.check_types(left_type, node.args[i].typ) {
+					base_arg_type := c.unwrap_generic(node.args[i].typ)
+					if c.check_types(base_arg_type, info.elem_type) {
+						if !base_arg_type.is_ptr() && info.elem_type.is_ptr()
+							&& info.elem_type.share() == .mut_t {
+							c.error('cannot ${method_name} `${arg_sym.name}` to `${left_sym.name}`',
+								arg.expr.pos())
+							continue
+						}
+					} else if !c.check_types(base_arg_type, c.unwrap_generic(left_type)) {
 						c.error('cannot ${method_name} `${arg_sym.name}` to `${left_sym.name}`',
 							arg.expr.pos())
 						continue

--- a/vlib/v/checker/tests/array_of_refs_insert_non_ref.out
+++ b/vlib/v/checker/tests/array_of_refs_insert_non_ref.out
@@ -1,0 +1,13 @@
+vlib/v/checker/tests/array_of_refs_insert_non_ref.vv:16:12: error: cannot append `Dummy` to `[]&Dummy`
+   14 |         v: 1
+   15 |     }
+   16 |     c.vals << d
+      |               ^
+   17 |     c.vals.insert(0, d)
+   18 | }
+vlib/v/checker/tests/array_of_refs_insert_non_ref.vv:17:19: error: cannot insert `Dummy` to `[]&Dummy`
+   15 |     }
+   16 |     c.vals << d
+   17 |     c.vals.insert(0, d)
+      |                      ^
+   18 | }

--- a/vlib/v/checker/tests/array_of_refs_insert_non_ref.vv
+++ b/vlib/v/checker/tests/array_of_refs_insert_non_ref.vv
@@ -1,0 +1,18 @@
+struct Dummy {
+	v int
+}
+
+@[heap]
+struct Container {
+mut:
+	vals []&Dummy
+}
+
+fn main() {
+	mut c := &Container{}
+	d := Dummy{
+		v: 1
+	}
+	c.vals << d
+	c.vals.insert(0, d)
+}

